10.2/healthcheck.sh

```diff
@@ -0,0 +1,296 @@
+#!/bin/bash
+#
+# Healthcheck script for MariaDB
+#
+# Runs various tests on the MariaDB server to check its health. Pass the tests
+# to run as arguments. If all tests succeed, the server is considered healthy,
+# otherwise it's not.
+#
+# Arguments are processed in strict order. Set replication_* options before
+# the --replication option. This allows a different set of replication checks
+# on different connections.
+#
+# --su{=|mysql} is option to run the healthcheck as a different unix user.
+# Useful if mysql@localhost user exists with unix socket authentication
+# Using this option disregards previous options/tests, so should usually be the
+# first option.
+#
+# Some tests require SQL privileges.
+#
+# TEST                      GRANTS REQUIRED
+# connect                   none*
+# innodb_buffer_pool_loaded USAGE
+# galera_online             USAGE
+# replication               SUPER or REPLICATION_CLIENT or REPLICA MONITOR (10.5+)
+# mariadbupgrade            none, however unix user permissions on datadir
+#
+# The SQL user used is the default for the mysql client. This can be the unix user
+# if no user(or password) is set in the [mariadb-client] section of a configuration
+# file. --defaults-{file,extra-file,group-suffix} can specify a file/configuration
+# different from elsewhere.
+#
+# Note * though denied error message will result in error log without
+#      any permissions.
+
+set -eo pipefail
+
+_process_sql()
+{
+	mysql ${nodefaults:+--no-defaults} \
+		${def['file']:+--defaults-file=${def['file']}} \
+		${def['extra_file']:+--defaults-extra-file=${def['extra_file']}} \
+		${def['group_suffix']:+--defaults-group-suffix=${def['group_suffix']}} \
+		-B "$@"
+}
+
+# TESTS
+
+
+# CONNECT
+#
+# Tests that a connection can be made over TCP, the final state
+# of the entrypoint and is listening. The authentication used
+# isn't tested.
+connect()
+{
+	set +e +o pipefail
+	mysql ${nodefaults:+--no-defaults} \
+		${def['file']:+--defaults-file=${def['file']}} \
+		${def['extra_file']:+--defaults-extra-file=${def['extra_file']}}  \
+		${def['group_suffix']:+--defaults-group-suffix=${def['group_suffix']}}  \
+		-h localhost --protocol tcp -e 'select 1' 2>&1 \
+		| grep -F "Can't connect" > /dev/null
+	local ret=${PIPESTATUS[1]}
+	set -eo pipefail
+	if [ "$ret" -eq 0 ]; then
```

Comment:
I personally prefer the use of `(( ... ))` for numerical comparison.
```bash
if (( ret == 0 )); then
if (( ret >= 0 )); then
```

Reason (again, very personal): `-gt` or `-eq` are specific to shell/bash and this makes it more difficult to read for non-bash programmers.

Reply:
nice, I missed totally what `((` was for. Now I know.